base/compiler/ssair/passes.jl

```diff
@@ -1098,34 +1098,44 @@ function try_inline_finalizer!(ir::IRCode, argexprs::Vector{Any}, idx::Int, mi::
     return true
 end
 
-is_nothrow(ir::IRCode, pc::Int) = ir.stmts[pc][:flag] & (IR_FLAG_EFFECT_FREE | IR_FLAG_NOTHROW) ≠ 0
+is_nothrow(ir::IRCode, pc::Int) = ir.stmts[pc][:flag] & (IR_FLAG_NOTHROW | IR_FLAG_EFFECT_FREE) ≠ 0
 
-function try_resolve_finalizer!(ir::IRCode, idx::Int, finalizer_idx::Int, defuse::SSADefUse, inlining::InliningState)
-    # For now: Require that all uses and defs are in the same basic block,
-    # so that live range calculations are easy.
-    bb = ir.cfg.blocks[block_for_inst(ir.cfg, first(defuse.uses).idx)]
+function try_resolve_finalizer!(ir::IRCode, idx::Int, finalizer_idx::Int, defuse::SSADefUse, inlining::InliningState, lazydomtree::LazyDomtree)
+    # For now: Require that all the uses and defs are dominated by `finalizer` block
+    # TODO: further relax this restriction by allowing a case when `finalizer` block is post-dominated by them
     minval::Int = typemax(Int)
     maxval::Int = 0
-
-    function check_in_range(x::Union{Int,SSAUse})
-        if isa(x, SSAUse)
-            didx = x.idx
+    function update_idx!(duidx::Int)
+        if duidx < minval
+            minval = duidx
+        end
+        if duidx > maxval
+            maxval = duidx
+        end
+    end
+
+    domtree = get!(lazydomtree)
+    finalizer_bb = block_for_inst(ir, finalizer_idx)
+    update_idx!(finalizer_idx)
+    alloc_bb = block_for_inst(ir, idx)
+    dominates(domtree, alloc_bb, finalizer_bb) || return nothing
+    update_idx!(idx)
+    function check_defuse(x::Union{Int,SSAUse})
+        duidx = x isa SSAUse ? x.idx : x
+        duidx == finalizer_idx && return true
+        bb = block_for_inst(ir, duidx)
+        if dominates(domtree, finalizer_bb, bb)
+            update_idx!(duidx)
+            return true
+        # TODO elseif post_dominates(domtree, bb, finalizer_bb)
+        #     update_idx!(duidx)
+        #     return true
         else
-            didx = x
-        end
-        didx in bb.stmts || return false
-        if didx < minval
-            minval = didx
+            return false
         end
-        if didx > maxval
-            maxval = didx
-        end
-        return true
     end
-
-    check_in_range(idx) || return nothing
-    all(check_in_range, defuse.uses) || return nothing
-    all(check_in_range, defuse.defs) || return nothing
+    all(check_defuse, defuse.uses) || return nothing
+    all(check_defuse, defuse.defs) || return nothing
 
     # For now: Require all statements in the basic block range to be nothrow.
     all(minval:maxval) do idx::Int
```

Comment:
This should probably only look at the statements in blocks on the domination path, but we can do that in the followup. I'd also like a `error_is_fatal` optimizer setting that assumes that any errors are fatal and thus finalizers don't need to run after errors, implying that we can skip this legality check.